Announcement blog post (#291)
nodejs/modules: doc/announcement.md

We’re making this change to align with browsers, where [extensions are required](https://github.com/WICG/import-maps#extension-less-imports). Part of the reason for Node.js to support ES modules in the first place is because ES modules are the standard, and the point of aligning to standards is to increase compatibility and interoperability between systems. In this case, Node.js supporting ES modules means that more JavaScript code will be capable of running in either Node or in browsers, without requiring some kind of translation process to make the code compatible with one or the other. Of course, it will always remain possible for users to write code that can’t function cross-platform (JavaScript referencing `window`, or importing `fs`, for example) but by aligning as much as we can, we encourage code to be cross-platform by default.

> Also, [import maps may bring support for similar functionality to browsers](https://github.com/WICG/import-maps#extension-less-imports). Should that proposal be accepted, Node is interested in supporting extensionless specifiers in a cross-platform compatible way.

Reviewer: We mention (and link to) import maps, like, four times in this post. Can we consolidate/reorganize all those bits to they're all together/unified (at least having the one call out three paragraphs below and then the one before this one and this one all seem slightly repetitive).

Author: This is a fair point, but I’m not sure how to address it. The references are all very different:

- In the introduction, import maps are briefly mentioned as part of the context of ESM arriving in browsers.
- In the section on extension searching, they’re mentioned as a way to achieve “extensionless” specifiers in browsers.
- In the final “Works in progress” section, they’re mentioned as similar to our “package path maps” feature.

That’s it. I suppose we could remove the last mention entirely, though I feel it’s relevant; but the other two feel important to keep. How would you suggest changing things?

Reviewer: Keep the sections, as is, remove the justifications and likening to import maps, put them into their own section after all the features have been introduced, maybe?

Author: > Keep the sections, as is, remove the justifications and likening to import maps, put them into their own section after all the features have been introduced, maybe?

You mean don’t mention import maps at all in any of the three sections listed above? Why not? I think they really should be mentioned at least in the intro paragraph about ESM in browsers.

Import maps isn’t our proposal and we’re not implementing any of it in Node, so I feel like it would be quite odd for it to get its own section.

I can see import maps as not being essential to mention in the two feature sections, if you think those mentions detract. Would you still want those mentions removed even if import maps doesn’t get its own section?

Reviewer: We won't actually be _implementing_ import maps though - in no way do they indicate a feature on their own. They are merely some work we're keeping in mind as we work. We'd be free to come to the same feature set without import maps existing if we so desired. I just think that if you think the justifications are for some reason warranted, rather than repeating it three times, doing it all at once in a single section at the bottom would scan much better.

> # Announcing a New `--experimental-modules`

Reviewer 2: nit: the `New` here should be `new` (all lowercase)

Back in 2017, Node.js 8.9.0 shipped experimental support for [ECMAScript modules](https://hacks.mozilla.org/2018/03/es-modules-a-cartoon-deep-dive/), known for their `import` and `export` statements. This support was behind the flag `--experimental-modules`.

A lot has happened since then. All major browsers [now support](https://caniuse.com/#feat=es6-module) ECMAScript modules (ES modules) via `<script type="module">`. Various projects have sprung up to make npm packages with ES module sources available for use in browsers via `<script type="module">`. Support for [import maps](https://github.com/WICG/import-maps), which bring to browsers Node-style package names in `import` statements, [is coming to Chrome](https://groups.google.com/a/chromium.org/forum/#!msg/blink-dev/qYeQFqgFOyA/rXJapjMaEAAJ).

> All this progress toward adoption of ES modules has increased the urgency for Node.js to ship its support for ES modules. Now that there are other runtimes and environments where ES modules are in use, it’s more important than ever that Node.js support the JavaScript standard.

Reviewer 3: nit - perhaps "**this** JavaScript standard" to be clear?

> Node’s initial ES modules support remained in an experimental state in order to allow the community time to provide feedback on that design. The [Modules Team](https://github.com/nodejs/modules) was formed to act on this feedback and ship first-class support for ES modules in Node.js. This has led to a [new implementation](https://github.com/nodejs/ecmascript-modules) for supporting ES modules, that we are pleased to announce will ship as part of Node.js 12. It will replace the old `--experimental-modules` implementation, behind the same flag. We hope that this new implementation addresses many of the community’s concerns, and can ship as part of Node.js proper, without a flag, before Node.js 12 reaches [LTS status](https://nodejs.org/en/about/releases/) in October 2019.

Reviewer 3: Node's -> Node.js's. I believe the policy is never to use "Node" without the ".js".

Author: I thought that only applied to documentation. Does it apply to all official communications?

- [`export` statements](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Statements/export) in ES module files can specify both default and named exports for `import` statements to reference.

- Dynamic `import()` expressions can be used to import ES modules from either CommonJS or ES module files. Note that `import()` returns a promise.

- `import.meta.url` provides the `file://` URL of the current ES module file.

- Loaders can be written to modify Node.js’s runtime behavior with respect to ES modules. _This is still very much a work in progress._

- Node.js can be run with an ES module file as a program’s initial entry point.

- Files loaded as ES modules are loaded in strict mode, which in CommonJS requires adding `'use strict';` to the top of every file.

- Files ending in `.mjs` are explicitly treated as ES modules in `import` statements and when run via the `node` command.

And the new version of `--experimental-modules` adds:

## `import` and `export` syntax in `.js` files

> We heard some [very](https://github.com/dherman/defense-of-dot-js/blob/master/proposal.md) [strong](https://gist.github.com/ceejbot/b49f8789b2ab6b09548ccb72813a1054) [feedback](https://twitter.com/maybekatz/status/1062473765865512961) that Node.js needs to provide a way to use `import` and `export` syntax in `.js` files.

Reviewer 4: Thanks, this is much better - I'd still prefer to indicate that this was always part of the plan (just was never planned to be the default, as it still is not).

Author: This language was the result of a lot of back and forth between various people, so I'd rather not open it up again for debate if you don't mind? Could this be acceptable to you?

Reviewer 4: I'm not sure why "how many people that weren't me" you talked to should mean that I don't get to retain my opinion here.

Author: Looking at the previous implementation’s [user-facing docs](https://nodejs.org/api/esm.html) and [design doc](https://github.com/nodejs/node-eps/blob/master/002-es-modules.md), the previous plan for supporting ESM in `.js` files was via a loader as shown in the example [here](https://nodejs.org/api/esm.html#esm_resolve_hook).

How about we change “Node.js needs to provide a way to use `import` and `export` syntax in `.js` files” to “Node.js needs to provide a way to use `import` and `export` syntax in `.js` files without requiring a loader.” Would that be acceptable?

cc @weswigham @robpalme

Reviewer 5: Guy and myself had plenty of talks about different ways to declare formats prior to the while forming, if you need me to dig up gists etc. I can, but the talks did occur both in EP and after the initial implementation. I am rather busy but is it really necessary to bicker about the timelines so much?

Reviewer 6: My original point was that the history doesn't matter. 

Whilst I'm happy with the wording as-is, if you wish to discard the historical/controversy angle, that's fine too. The content already sells itself, so the more boring/factual the better IMO.

Reviewer 4: per offline discussion:
```suggestion
We heard feedback that Node.js needs to provide a way to use `import` and `export` syntax in `.js` files.
```

Reviewer 7: It may be worthwhile to note that feedback is different from what was heard, imho wording this needs to balance from both sides and be slightly biased towards the reader (ie they will be doing the effort to read and use this after all).

```suggestion
We are make a lot of progress towards popular features demanded by members of the community and in ways that will ensure that all existing projects can continue to operate smoothly.

With this new implementation, we are making it possible for the first time to actually `import` and `export` syntax in `.js` files with opt-in package configuration or CLI flags directly, ie no longer requiring a custom loader.
```

This suggestion focuses on the substance, **but** I sincerely think that references from which we claim to infer such demand are necessary somewhere in the announcement — it is appropriate blogging decorum and ensure tracking with a proper chronological structure.

So we need a way to relate to those links:

- https://github.com/dherman/defense-of-dot-js/blob/master/proposal.md
- https://gist.github.com/ceejbot/b49f8789b2ab6b09548ccb72813a1054
- https://twitter.com/maybekatz/status/1062473765865512961

I suggest considering a second statement in the first paragraph making the case from those links that is addressed in the next paragraph. This would flow well for the reader and be very balanced to both Node.js and the community.

I worry that being too sensitive here comes off negatively for Node.js — did we not just give a solution that was asked for here after insisting we will never ever do that?! If we tiptoe about that, the reader will be inclined to think it, but if we assert the correct narrative, it keeps readers focused on the exciting new features, and frames those links as one community working towards one goal.

Reviewer 7: That suggestion (which I strongly recommend would look more like this):

```suggestion
We are make a lot of progress towards some [popular features demanded by members of the community](https://twitter.com/maybekatz/status/1062473765865512961) and in ways that will ensure that all existing projects can continue to operate smoothly. This required a lot of careful consideration and insights from community discussions on how [the community uses ESM in projects today](https://github.com/dherman/defense-of-dot-js/blob/master/proposal.md) and  [possible ways for packages to move forward](https://gist.github.com/ceejbot/b49f8789b2ab6b09548ccb72813a1054).

With this new implementation, we are making it possible for the first time to actually `import` and `export` from `.js` files with opt-in package configuration or CLI flags directly, ie no longer requiring a custom loader.
```

<details><summary>Preview</summary><blockquote>

We are make a lot of progress towards some [popular features demanded by members of the community](https://twitter.com/maybekatz/status/1062473765865512961) and in ways that will ensure that all existing projects can continue to operate smoothly. This required a lot of careful consideration and insights from community discussions on how [the community uses ESM in projects today](https://github.com/dherman/defense-of-dot-js/blob/master/proposal.md) and  [possible ways for packages to move forward](https://gist.github.com/ceejbot/b49f8789b2ab6b09548ccb72813a1054).

With this new implementation, we are making it possible for the first time to actually `import` and `export` from `.js` files with opt-in package configuration or CLI flags directly, ie no longer requiring a custom loader.

</blockquote></details>

Reviewer 4: No, because nobody insisted we would never do that. With or without any of this feedback, we would always have provided some way to have ESM in .js - predating the WG as well. Thus, the links aren’t relevant, and would further the misconception that this is a newly intended feature.

Reviewer 7: @ljharb can you clarify what the intent was:

1. Allow a `package.json` to make it possible to `node main.js` to be ESM: Yes/Loader/No
2. Allow a flag `node --flag main.js` to be ESM: Yes/Loader/No
3. Allow a package to be published with `.js` files as ESM: Yes/Loader/No
4. Allow a package to be published with `.js` files as ESM & CJS: Yes/Loader/No

Can you reply on each one to help me get a more complete picture please.

Because from the end user's side (which I was at roughly 12 months ago) things looked a certain way, and regardless of what the intent was, it is important to recognize that no matter how hard we all try, the other perspective is always very hard fully appreciate — and naturally mine was one of a the wide spectrum of perspectives, but not an outlier

**As such**, it is important for us to be as objective and thorough as possible when talking to the audience.

Reviewer 4: @SMotaal 
1. afaik undiscussed, but i don't see any reason why we wouldn't have - if package.json has any control then it should have entry point control.
1. yes, since day one, if for nothing else for stdin/eval/extensionless files.
1. an open question, but leaning towards yes because if it can work for non-bare imports via package.json (ie, for your own top-level files) than it wouldn't make sense to actively prohibit it for published files.
1. dual packages have always been intended; without that as an option, there's no migration strategy.

It is totally valid that some user perspectives may have been solely informed from the implementation; it's also valid that user perspectives *do not get to dictate* what the intention was.

In general, being objective also means not validating "internet mob outrage made change happen" so as to ensure that's not seen as an effective means to effect change in the future.

Reviewer 7: Thanks for clarifying

> The new `--experimental-modules` provides two ways: a `"type"` field in `package.json` and a `--type` flag. Their naming was inspired by browsers’ `<script type="module">`.

Reviewer 3: ```suggestion
The new `--experimental-modules` provides two ways to achieve this: a `"type"` field in `package.json` and a `--type` flag for the entry point. Their naming was inspired by browsers’ `<script type="module">`.
```

Perhaps also add:

```
Ideally setting this flag can become part of the package manager init process.
```

or something to that affect?

Author: A _lot_ of back-and-forth went into this section, I don’t want to upset any compromises by changing it further if you don’t mind.

Reviewer 3: The changes are purely stylistic edits, as opposed to content.

### `package.json` `"type"` field

Add `"type": "module"` to the `package.json` for your project, and Node will treat all `.js` files in your project as ES modules.

If some of your project’s files use CommonJS and you can’t convert your entire project all at once, you can either rename those files to use the `.cjs` extension or put them in a subfolder containing a `package.json` with `{ "type": "commonjs" }`, under which all `.js` files are treated as CommonJS.

For any file that Node.js tries to load, it will look for a `package.json` in that file’s folder, then that file’s parent folder and so on upwards until it reaches the root of the volume. This is similar to [how Babel searches for `.babelrc` files](https://babeljs.io/docs/en/config-files#file-relative-configuration). This new approach allows Node.js to use `package.json` for package-level metadata and configuration, similar to how it is already used by Babel and other tools.

### `--type` flag

> Use `--type=module`, or the shorthand `-m`, to run a `.js` file as an ES module.

Reviewer 3: Perhaps be clear about the full usage here in the example - `node --experimental-modules --type=module main.js`?

## `import` for JavaScript only

The previous `--experimental-modules` allowed `import` statements of JSON and native modules. This has been removed; you may use `module.createRequireFromPath()` for these.

A separate flag `--experimental-json-modules` enables experimental support for `import` of JSON files. There is [work in progress](https://github.com/whatwg/html/issues/4315) for standardizing this feature with browsers, and Node hopes to align our support with the eventual standard.

There is also ongoing work to cover WASM and other future potential module types. Node.js will add support for these in spec-compliant ways over time.

## ES module code in packages

_This is a work in progress and subject to change._ You can create a package with ES module sources by using the `package.json` `"main"` field to point to an ES module package entry point. Node will know to load it as an ES module if the file ends in `.mjs` or if the `package.json` also contains `"type": "module"`.

> Currently it is not possible to create a package that is usable both via `require` (as CommonJS) and via `import` (as an ES module). There are efforts underway to address this, and may involve changes to the above (especially concerning `"main"`). Please do not publish any ES module packages intended for use by Node.js until this is resolved.

Reviewer 3: I'm not sure how genuine this paragraph is.

Perhaps be more explicit - "it is not possible to have a package main that is usable by both `require` and `import`"?

Even that isn't quite right as CommonJS is perfectly usable by import.

Perhaps the exact version is "it is not currently possible to have a package main that exposes an ES module for `import` and a CommonJS module for `require`"?

Then it would be worth using the term "dual mode" here as something that is being explored.

Author: > Perhaps the exact version is “it is not currently possible to have a package main that exposes an ES module for `import` and a CommonJS module for `require`“?

How is this meaningfully different than “Currently it is not possible to create a package that is usable both via `require` (as CommonJS) and via `import` (as an ES module).”? I didn’t want to mention “main” specifically here in case the ESM entry point ends up in a different field.

Reviewer 3: Well you can create a package that exposes both CommonJS and ES Modules just fine through `require('pkg/main.js')` and `import 'pkg/main.mjs'`. So this is specifically a main thing. I don't think there is any confusion in mentioning main by name, and not doing that makes this sound like we are now the package police.

Author: I want to join the package police. Where do I sign up?

I think I found a phrasing that avoids the issue that I found confusing where you mean “main” to be “package entry point” but I immediately think of it as meaning the `"main"` field (which defines the same thing, of course, but potentially only for CommonJS). How about this:

> Currently it is not possible to create a package that can be used via both `require('pkg')` and `import 'pkg'`.

That avoids needing the word “main” and it makes clear we’re talking about the bare specifier/package entry point.

Reviewer 3: > I want to join the package police. Where do I sign up?

You already did :)

> Please do not publish any ES module packages intended for use by Node.js until this is resolved.

Just to be really clear on this sentence, because I'm not sure what it means for me as a library author, are we effectively saying: "Be aware that publishing packages that rely on the "main" being a module should not be published because the exact semantics of ES modules are still changing and thus that package could break in future"?

I think it would help to be really really clear on what the line is for users here.

To try and spell it out, here is my best guess:

It is NOT ok to:

* Publish packages with "type": "module"
* Publish packages with "main": "main.mjs" being a module

as the above may break as the semantics change under --experimental-modules.

But it IS ok to:

* Publish packages that contain ES modules for consumption by build tools
* Publish packages with "module": "main.js" as this will likely not be supported
* Publish packages with require('esm')

Does that cover what you mean? It would help to understand.

Author: That covers what I mean, though I'm not warning against `type: "module"` because that's not on our roadmap as something lacking consensus or that might change in phase 3. But everything else, yes.

If you want to draft alternative language I'm happy to change this. One note I got from Wes though was to keep it short, so I guess somehow get all that across without the two bulleted lists?

## Works in progress

All of the above is shipping as part of `--experimental-modules` in Node.js 12. On our [road map](https://github.com/nodejs/modules/blob/master/doc/plan-for-new-modules-implementation.md) for potential improvements before the `--experimental-modules` flag is hopefully dropped in October 2019, when Node.js 12 reaches LTS status:

> - **Better loaders.** The current loaders code is a holdover from the previous `--experimental-modules` and doesn’t fully integrate with the new system. There is much work to be done here.

Reviewer 3: Can we rather have:

```
**Loaders features.** Loaders are being further developed to support process isolation, multiple loaders, multi-Realm support with lower-level hooks. The `--loader` API will still change considerably before this is unflagged.
```

or something along those lines.

> - **Dual CommonJS/ES module packages.** We want to provide a standard way for package authors to publish a package that can be both `require`d into CommonJS or `import`ed into an ES module.

Reviewer 3: Perhaps `We're investigating ways for` instead of "We want to provide"? That is, this something we're looking into as discussed previously.

Author: I think our discussions are pretty firm in the camp of us wanting to make this happen one way or another, and it’s on our roadmap. Several people expressed concerns about fracturing the npm registry if we don’t provide a way for dual packages to happen. I think we can be as forthright as “We want to provide”.

Reviewer 3: Then `We would like to provide` perhaps rather? I'm just hesitant to make future promises.

# Announcing a New `--experimental-modules`

Back in 2017, Node.js 8.9.0 shipped experimental support for [ECMAScript modules](https://hacks.mozilla.org/2018/03/es-modules-a-cartoon-deep-dive/), known for their `import` and `export` statements. This support was behind the flag `--experimental-modules`.

> A lot has happened since then. All major browsers [now support](https://caniuse.com/#feat=es6-module) ECMAScript modules (ES modules) via `<script type="module">`. Various projects have sprung up to make NPM packages with ES module sources available for use in browsers via `<script type="module">`. Support for [import maps](https://github.com/WICG/import-maps), which bring to browsers Node-style package names in `import` statements, [is coming to Chrome](https://groups.google.com/a/chromium.org/forum/#!msg/blink-dev/qYeQFqgFOyA/rXJapjMaEAAJ).

Reviewer 4: always lowercase npm everywhere; "ES module" → "ES Module"; "node" is lowercase unless it's "Node.js", i think?
```suggestion
A lot has happened since then. All major browsers [now support](https://caniuse.com/#feat=es6-module) ECMAScript modules (ES modules) via `<script type="module">`. Various projects have sprung up to make npm packages with ES Module sources available for use in browsers via `<script type="module">`. Support for [import maps](https://github.com/WICG/import-maps), which bring to browsers node-style package names in `import` statements, [is coming to Chrome](https://groups.google.com/a/chromium.org/forum/#!msg/blink-dev/qYeQFqgFOyA/rXJapjMaEAAJ).
```

> All this progress toward adoption of ES modules has increased the urgency for Node.js to ship its support for ES modules. Especially now that there are other runtimes and environments where ES modules are in use, it’s more important than ever that Node.js join the JavaScript standard.

Reviewer 4: ```suggestion
All this progress toward adoption of ES Modules has increased the urgency for Node.js to ship its support for ES Modules. Especially now that there are other runtimes and environments where ES Modules are in use, it’s more important than ever that Node.js join the JavaScript standard.
```

Author: I hear what you’re saying. I went with a version that tries not to have Node toot its own horn so much 😄 In this case, though, Node really is following.

> Yet Node’s initial ES modules support has remained in an experimental state, largely due to strong pushback from the community to several decisions that were made in its design. In response, the [Modules Team](https://github.com/nodejs/modules) was formed to take a fresh look at implementing support for ES modules in Node.js. This has led to a [new implementation](https://github.com/nodejs/ecmascript-modules) for supporting ES modules, that we are pleased to announce will ship as part of Node.js 12. It will replace the old `--experimental-modules` implementation, behind the same flag. We hope that this new implementation addresses many of the community’s concerns, and can ship as part of Node.js proper, without a flag, before Node.js 12 reaches [LTS status](https://nodejs.org/en/about/releases/) in October 2019.

Reviewer 4: ```suggestion
Yet node’s initial ES Modules support has remained in an experimental state, largely due to strong pushback from the community to several decisions that were made in its design. In response, the [Modules Team](https://github.com/nodejs/modules) was formed to take a fresh look at implementing support for ES Modules in Node.js. This has led to a [new implementation](https://github.com/nodejs/ecmascript-modules) for supporting ES Modules, that we are pleased to announce will ship as part of Node.js 12. It will replace the old `--experimental-modules` implementation, behind the same flag. We hope that this new implementation addresses many of the community’s concerns, and can ship as part of Node.js proper, without a flag, before Node.js 12 reaches [LTS status](https://nodejs.org/en/about/releases/) in October 2019.
```

Reviewer 4: > largely due to strong pushback from the community

I don't think this is entirely accurate. Certainly there was pushback, but I don't believe that's the reason it remained experimental prior to the WG getting spun up. I think this part should be severely tweaked or removed.

(Especially considering ".mjs by default" is one of the things that got the largest pushback, and that has *not* changed)

## What’s in `--experimental-modules`

Like the previous version, this new `--experimental-modules` adds support to Node.js for the following:

> - [ES2015 `import` statements](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Statements/import) can reference JavaScript files that are themselves written using ES module syntax. Files can be referenced as relative URLs (`'./file.mjs'`), absolute `file://` URLs (`'file:///opt/app/file.mjs'`), package names (`'es-module-package'`) or paths within packages (`'es-module-package/lib/file.mjs'`).

Reviewer 4: ```suggestion
- [ES2015 `import` statements](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Statements/import) can reference JavaScript files that are themselves written using ES Module syntax. Files can be referenced as relative URLs (`'./file.mjs'`), absolute `file://` URLs (`'file:///opt/app/file.mjs'`), package names (`'es-module-package'`) or paths within packages (`'es-module-package/lib/file.mjs'`).
```

Reviewer 4: I don't think this is resolved yet.

> - `import` statements that reference ES module files can specify both the default export (`import _ from 'es-module-package'`) or named exports (`import { shuffle } from 'es-module-package'`). All Node.js built-in packages like `fs` and `path` support both default and named exports.

Reviewer 4: ```suggestion
- `import` statements that reference ES Module files can specify both the default export (`import _ from 'es-module-package'`) or named exports (`import { shuffle } from 'es-module-package'`). All Node.js built-in packages like `fs` and `path` support both default and named exports.
```

- `import` statements that reference CommonJS files (all current JavaScript code written for Node.js, [using `require` and `module.exports`](https://nodejs.org/api/modules.html)) can use the CommonJS default export (`import _ from 'commonjs-package'`) only. _This is a work in progress and may change in the future._

> - [`export` statements](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Statements/export) in ES module files can specify both default and named exports for `import` statements to reference.

Reviewer 4: ```suggestion
- [`export` statements](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Statements/export) in ES Module files can specify both default and named exports for `import` statements to reference.
```

> - Dynamic `import()` expressions can be used to import ES modules into either CommonJS or ES module files. Note that `import()` returns a promise.

Reviewer 4: ```suggestion
- Dynamic `import()` expressions can be used to import ES Modules into either CommonJS or ES Module files. Note that `import()` returns a Promise.
```

> - `import.meta.url` provides the `file://` URL of the current ES module file.

Reviewer 4: ```suggestion
- `import.meta.url` provides the `file://` URL of the current ES Module file.
```

> - Node.js can be run with an ES module file as a program’s initial entry point.

Reviewer 4: ```suggestion
- Node.js can be run with an ES Module file as a program’s initial entry point.
```

> - Loaders can be written to modify Node’s runtime behavior with respect to ES modules. _This is still very much a work in progress._

Reviewer 4: ```suggestion
- Loaders can be written to modify Node’s runtime behavior with respect to ES Modules. _This is still very much a work in progress._
```

So if that’s what’s carried over from the previous implementation, what’s new?

## `import` and `export` syntax in `.js` files

> Yes, we heard [you](https://github.com/dherman/defense-of-dot-js/blob/master/proposal.md). And [you](https://gist.github.com/ceejbot/b49f8789b2ab6b09548ccb72813a1054). And [you](https://twitter.com/maybekatz/status/1062473765865512961). You can now use `import` and `export` syntax in `.js` files.

Reviewer 4: This seems a bit disingenuous, since that wasn't prevented before - it just was not, and still is not, the default.

Reviewer 4: How is this something you can "now" do? This was something that was always planned; whether it happened to be shipped prior to the WG forming or not.

I do not want any implication that having an ESM `.js` file is suddenly an option because of this working group; we have merely preserved (and perhaps implemented) that option.

Author: The article is describing the new implementation as compared to the previous implementation. It’s not describing the new implementation’s _plan_ as compared to the possibly not fully implemented _plan_ for the previous implementation. I don’t think most people will know (or care) what the previous plan was, just what’s new in this version of the software versus the previous version.

Reviewer 4: Given the "we heard you" rhetoric at the beginning of the paragraph, this is absolutely implying that it wasn't part of the plan from day one, which it was. If you're talking about implementation differences and capabilities, then talking about "responding to feedback", especially flippantly implying that those of us involved prior to the WG a) weren't listening, or b) weren't planning on enabling this anyways, doesn't belong here.

Author: > this is absolutely implying that it wasn’t part of the plan from day one, which it was.

If supporting ESM in `.js` was ever the plan, shipping without any support for it certainly sent a message to the community that that wasn’t the final plan. I was an interested observer when the first `--experimental-modules` came out, probably a lot more interested than most, and I had no idea that ESM in `.js` was ever on Node’s roadmap.

Regardless, plans are never mentioned here. This text is just saying that we hear the request from the community loud and clear, and we’re responding to it. That’s not an indictment of anyone involved in the process so far. If anything, it shows that we’re good developers who are working hard to build the best product for our users.

Reviewer 4: I'm still not OK with the wording in this paragraph. I'm not sure how to explain it any clearer, but please stop marking it as resolved; it is not.

Reviewer: In terms of tone, I think it has much the same issue as the "we didn't do this to annoy you" line did. In isolation it might seem conciliatory, but mentioning it at all is only likely to incense anyone who feels these results are only a half-measure. It doesn't sound good to many of the people it's meant to ease the minds of.

Reviewer 6: All we need is a simple statement that modules can use the JS extension.

That will be enough. It's super clear. 

We should avoid referencing individuals that have taken particular positions in the past, because that frames it as a conflict between different sides, whereas it's more of a collaboration.

> The `.mjs` extension is still supported, and still always explicitly means that its file uses the Module parse goal (and therefore is an ES module). We also created a new `.cjs` file extension, to explicitly signify that a file should be treated as CommonJS. (CommonJS is the other module system that Node.js supports, with `require` and `module.exports`.) These two extensions allow individual files to define how they should be loaded, as ES modules or as CommonJS.

Reviewer 4: ```suggestion
The `.mjs` extension is still supported, and still always explicitly means that its file uses the Module parse goal (and therefore is an ES Module). We also created a new `.cjs` file extension, to explicitly signify that a file should be treated as CommonJS. (CommonJS is the other module system that Node.js supports, with `require` and `module.exports`.) These two extensions allow individual files to define how they should be loaded, as ES modules or as CommonJS.
```

Reviewer 4: I still object to the `.cjs` extension - I feel it was snuck in and not properly discussed in a meeting - and I do not want it mentioned *in any way* publicly, since I intend to have it removed.

Reviewer 4: #293

> As it turns out, a lot of people didn’t like needing to use a new file extension for JavaScript. Browsers don’t require it, tools that build and process JavaScript have issues with it, and languages that transpile into JavaScript such as TypeScript and CoffeeScript don’t necessarily know when to save their output as `.js` or as `.mjs`. So while `.mjs` and `.cjs` have use cases that only they can solve (ES module and CommonJS files side-by-side in the same folder, loose JavaScript files outside of any package) the modules group searched for solutions other than `.mjs` to let Node.js know when to interpret a file as an ES module.

Reviewer 4: Re OP: this entire paragraph.

Reviewer 6: Agree with @weswigham - let's not mention tools.  Anything we state there is likely to become outdated.  It's also a catch-22 situation, because part of the reason for tools not updating is that they are waiting for a clear signal from Node.  Blog posts like *this* ought to be that clear signal rather than muddying the waters.

For any file that Node.js tries to load, it will look for a `package.json` in that file’s folder, then that file’s parent folder and so on upwards until it reaches the root of the volume. This is similar to [how Babel searches for `.babelrc` files](https://babeljs.io/docs/en/config-files#file-relative-configuration). This new approach allows Node.js to use `package.json` for package-level metadata and configuration, similar to how it is already used by Babel and other tools.

### `--type` flag

Use `--type=module`, or the shorthand `-m`, to run a `.js` file as an ES module.

When running a file, e.g. `node file.js`, Node follows an algorithm to determine if it should load the file as CommonJS or as an ES module. First it looks for an explicit file extension (`.mjs` or `.cjs`); then it looks for a `"type"` field in the nearest parent `package.json`; and finally it looks at the `--type` flag. The `--type` flag can be `--type=module` (or `-m`) or `--type=commonjs`. You can also set this flag in `NODE_OPTIONS` if you’d like to change how Node behaves systemwide.

> This flag provides a way to support ES module syntax for input via `--eval`, `--print` or `STDIN`.

Reviewer 4: ```suggestion
This flag provides a way to support ES module syntax for input via `--eval`, `--print`, or `STDIN`.
```

> ## Mandatory file extensions

Reviewer 6: This section title is a bit ominous until you read the detail.  It kinda also suggests that the underlying filenames themselves must have extensions, which is untrue.  Could we rename it to "Explicit filenames"?

In the new `--experimental-modules`, file extensions are mandatory in `import` statements: `import './file.js'`, not `import './file'`. However, the CommonJS-style automatic extension resolution behavior (`'./file'`) can be enabled via a new flag, `--es-module-specifier-resolution=node`. (Its inverse, the default, is `--es-module-specifier-resolution=explicit`.) Package names are still just package names, e.g. `import fs from 'fs'`.

> This change isn’t because we’re trying to annoy an entirely new constituency now that `.mjs` is no longer required. We recognize that there are years’ worth of JavaScript out there written in the extensionless style, and this will feel like a huge breaking change for all that code. That’s why we’re providing the `--es-module-specifier-resolution=node` flag, and this is also subject to change before `--experimental-modules` itself ships unflagged.

Reviewer: ```suggestion
We’re providing the `--es-module-specifier-resolution=node` flag to opt in to cjs-style extension and `index` resolution. We've turned it off by default to collect feedback on how users feel about using fully specified paths before we unflag the implementation. You can find our discussion on the topic [here](https://github.com/nodejs/modules/issues/268).
```

Especially don't mention "we didn't do this just to annoy you", because that's just going to rile up the people it annoys. I should know, I'm one of them, and reading it riled me. It's like the younger sibling saying "I'm not touching you" while holding their finger an inch away, pointed at you. It might technically be true, but the annoying effect is potentially magnified.

In addition, linking to our discussion on the topic lets those who are interested read the relevant background, and elides it for those who are not (although the issue doesn't collate all the meeting minutes on the topic, and there's a lot of them).

> This change isn’t because we’re trying to annoy an entirely new constituency now that `.mjs` is no longer required. We recognize that there are years’ worth of JavaScript out there written in the extensionless style, and this will feel like a huge breaking change for all that code. That’s why we’re providing the `--es-module-specifier-resolution=node` flag, and this is also subject to change before `--experimental-modules` itself ships unflagged.

Reviewer 6: Agree with @weswigham. The real danger line for me is "this will feel like a huge breaking change for all that code" - we shouldn't be referring to "breaking changes" in an experimental release and we shouldn't presume how people will feel.  The following paragraph already does the job of focusing on the positives/rationale, so I think we can just omit this paragraph.

> We’re making this change to align with browsers, where [extensions are required](https://github.com/WICG/import-maps#extension-less-imports). Part of the reason for Node.js to support ES modules in the first place is because ES modules are the standard, and the point of aligning to standards is to increase compatibility and interoperability between systems. In this case, Node.js supporting ES modules means that more JavaScript code will be capable of running in either Node or in browsers, without requiring some kind of translation process to make the code compatible with one or the other. Of course, it will always remain possible for users to write code that can’t function cross-platform (JavaScript referencing `window`, or importing `fs`, for example) but by aligning as much as we can, we encourage code to be cross-platform by default.

Reviewer 4: this is incorrect; extensions are **not** required in browsers - browsers ignore extensions entirely and have no opinion on them. Browsers require an exact URL, and any resolution or normalization (as in any other request on the web) needs to be handled by the server. This includes adding an extension, rewriting the path, adding or removing trailing slashes, etc.

Author: From a user’s perspective, extensions are required. They need to type `import './file.js'`, not `import './file'`. I think that getting into the weeds of MIME types and server resolution is too technical for a broad audience document.

Reviewer 4: `import './file'` will work fine if the server can serve that URL. I'm not talking about MIME types.

Note that rewrite rules to remove extensions for pages (altho usually not for assets) are absurdly common in PHP apache/httpd configs, and in Rails, so this isn't something that's likely to confuse most users.

Author: > (altho usually not for assets)

Yes, and `.js` files are generally always assets. All the [examples](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Statements/import#Description) that I’ve ever seen of using `import` in a browser environment show the extension in the source code, so from the user’s perspective they need to type the extension.

We can change “extensions are required” to “extensions are (generally) required” to be more legalistically correct, but I think that just makes things all the more confusing as users will wonder when they might not be required, and I don’t think we should be discussing server-side rewrites here.

Reviewer 4: I don't think we can use the word "required" at all - browsers *do not care* about extensions.

> Also, [import maps may bring support for automatic extension resolution to browsers](https://github.com/WICG/import-maps#extension-less-imports). Should that proposal be accepted, Node is interested in supporting extensionless specifiers in a cross-platform compatible way.

Reviewer 4: this is also inaccurate; it won't be "automatic" extension resolution, it will merely be clientside resolution. "adding an extension" is something that has to be statically precomputed by the server (as opposed to the server dynamically handling it at request time using rewriting rules)

> We’re making this change to align with browsers, where [extensions are required](https://github.com/WICG/import-maps#extension-less-imports). Part of the reason for Node.js to support ES modules in the first place is because ES modules are the standard, and the point of aligning to standards is to increase compatibility and interoperability between systems. In this case, Node.js supporting ES modules means that more JavaScript code will be capable of running in either Node or in browsers, without requiring some kind of translation process to make the code compatible with one or the other. Of course, it will always remain possible for users to write code that can’t function cross-platform (JavaScript referencing `window`, or importing `fs`, for example) but by aligning as much as we can, we encourage code to be cross-platform by default.

Author: We should discuss this in the meeting. The point of requiring explicit extensions by default was to spark a discussion in the community, and part of that involves us explaining why that decision was made. I’m happy to tone this down if you think it’s too strong, but I don’t think we can just mention that filenames need extensions now with no context and no explanation.

When I mentioned not favoring any specific proposals, I meant for items not yet implemented.

> We’re making this change to align with browsers, where [extensions are required](https://github.com/WICG/import-maps#extension-less-imports). Part of the reason for Node.js to support ES modules in the first place is because ES modules are the standard, and the point of aligning to standards is to increase compatibility and interoperability between systems. In this case, Node.js supporting ES modules means that more JavaScript code will be capable of running in either Node or in browsers, without requiring some kind of translation process to make the code compatible with one or the other. Of course, it will always remain possible for users to write code that can’t function cross-platform (JavaScript referencing `window`, or importing `fs`, for example) but by aligning as much as we can, we encourage code to be cross-platform by default.

Reviewer: I don't think if you need to convince people with extra learnings that it's a great idea... That it's that great of an idea.

I would much rather people gave feedback at this stage without being colored by our arguments for or against first, since that visceral reaction is what's going to spread, eg, on Twitter, upon the final release anyway~

And as I said, linking to the discussion via clickthru seems fine - the entire discussion is just too heavy for the announcement.

> Also, [import maps may bring support for automatic extension resolution to browsers](https://github.com/WICG/import-maps#extension-less-imports). It would be premature for Node to implement its own solution to supporting extensionless specifiers while efforts are underway to provide a cross-platform standard. Again, this is subject to change.

Reviewer 6: Could we phrase this in a more collaborative way? It should be Node collaborating and tracking this proposal rather than Node being blocked by it.

## `module.createRequireFromPath`

The “CommonJS globals” `require`, `exports`, `module.exports`, `__filename`, `__dirname` are not defined in ES modules. However, [`module.createRequireFromPath()`](https://nodejs.org/api/modules.html#modules_module_createrequirefrompath_filename) can be used to create a CommonJS `require` function to be used in an ES module context.

## `import` for JavaScript only

The previous `--experimental-modules` allowed `import` statements of JSON and native modules. This has been removed; you may use `module.createRequireFromPath()` for these.

> There is ongoing work in the standards space to cover JSON and WASM and other future potential module types. Node.js will add support for these in compliant ways when the standard is finalized.

Reviewer 4: We should link to #255 and https://github.com/whatwg/html/issues/4315 here, and indicate that it's likely something we'll add in the future. There *is* no standard around importing JSON that applies to node, nor will there be - we'd be adding it unilaterally (with the intent of matching browsers) - so there's no finalization to wait for.

> The “CommonJS globals” `require`, `exports`, `module.exports`, `__filename`, `__dirname` are undefined in ES modules. However, [`module.createRequireFromPath()`](https://nodejs.org/api/modules.html#modules_module_createrequirefrompath_filename) can be used to create a CommonJS `require` function to be used in an ES module context.

Reviewer 6: "undefined" -> "not defined"

## ES module code in packages

_This is a work in progress and subject to change._ You can create a package with ES module sources by using the `package.json` `"main"` field to point to an ES module package entry point. Node will know to load it as an ES module if the file ends in `.mjs` or if the `package.json` also contains `"type": "module"`.

Currently it is not possible to create a package that is usable both via `require` (as CommonJS) and via `import` (as an ES module). There are efforts underway to address this, and may involve changes to the above (especially concerning `"main"`). Please do not publish any ES module packages intended for use by Node.js until this is resolved.

> ## Coming soon

Reviewer 4: ```suggestion
## Possibly coming soon
```
given that not all these items have consensus, maybe best not to imply that they'll all actually happen

> Currently it is not possible to create a package that is usable both via `require` (as CommonJS) and via `import` (as an ES module). There are efforts underway to address this, and may involve changes to the above (especially concerning `"main"`). Please do not publish any ES module packages intended for use by Node.js until this is resolved.

Reviewer 8: Given this, do we want to talk about the details here, e.g. even mention using `main` above?

Reviewer 9: i think so. we want people to use these features, just not to depend on them in published packages.

Reviewer 8: Right, but this feature is specifically about published packages. So what would "using them" mean in this case if not using it in a published package?

Reviewer 9: you can use a package.json without publishing to npm :shrug:

Reviewer 8: True, but I'm not sure how common that is. My impression so far is that people are actively surprised when they are told that you *can* use `package.json` to influence file resolution without involving installed (or linked) packages.

Author: > Given this, do we want to talk about the details here

I wanted to shy away mentioning the proposals under consideration, to avoid seeming to prefer any of them (especially if we end up going with something not yet proposed) and because that felt like too much detail. I thought it was worth mentioning `"main"` specifically to tell users that the UX they will interact with might change, that we’re not just changing implementation details under the hood.

I’m very open to changing this paragraph if you’d like to suggest something better, I was just trying to prevent another `"module"` situation where the community started relying on something that wasn’t standardized.

Reviewer 3: It could also be worth mentioning "module" here, and why we may not be able to support it at all. So that those doubling down on this can be made aware.

Reviewer 7: Distinctions to be made here about publishing packages:

1. Not published and simply used locally for **non-production** testing — *highly encouraged*
2. Published to be used internally and strictly for **non-production** testing when warranted — *not recommended*
3. Published first time for public use as a third-party dependency — *discouraged*
4. Published as an update of the same — *highly discouraged*

Some people might need a little more to go on with respect to when packages are intended for Node.js and/or browser venues where they would want to ensure their ESM code is handled correctly making its way through those channels too. So I don't think "Please do not publish any ES module packages intended for use by Node.js until this is resolved." gives enough in way of what people should be experimenting with and what they should completely avoid, imho.

Reviewer: > especially concerning `"main"`

That's just, like, your opinion maaaaaaan. But moreso: The extra call-out doesn't really add anything to the sentence.

> Currently it is not possible to create a package that is usable both via `require` (as CommonJS) and via `import` (as an ES module). There are efforts underway to address this, and may involve changes to the above (especially concerning `"main"`). Please do not publish any ES module packages intended for use by Node.js until this is resolved.

Reviewer 4: it would be better to discourage proliferation of patterns in the ecosystem prior to unflagging; it makes eventual migration to One True Pattern much harder for the community.

All of the above is shipping as part of `--experimental-modules` in Node.js 12. On our [road map](https://github.com/nodejs/modules/blob/master/doc/plan-for-new-modules-implementation.md) for improvements before the `--experimental-modules` flag is hopefully dropped in October 2019, when Node.js 12 reaches LTS status:

- **Better loaders.** The current loaders code is a holdover from the previous `--experimental-modules` and doesn’t fully integrate with the new system. There is much work to be done here.

- **Dual CommonJS/ES module packages.** We want to provide a standard way for package authors to publish a package that can be both `require`d into CommonJS or `import`ed into an ES module.

> - **Easier `require`.** Using `module.createRequireFromPath()` involves a lot of boilerplate. We hope to provide a simpler way to use `require` in ES module code.

Reviewer 4: I'm not sure this is necessary if `import` can be used for CJS. Can we indicate that?

Author: This is just copied from the roadmap, so I wouldn’t make changes here that deviate from the roadmap.

To answer your direct point, though, as long as `require` can do things that `import` can’t (like include native modules, which we aren’t currently planning on supporting via `import`) then there will always be a need for `require` and `createRequireFromPath`, and so there will be a desire for getting `require` more easily. Maybe we never build it, but at the moment it’s on the roadmap.

> - **Dual CommonJS/ES module packages.** We want to provide a standard way for package authors to publish a package that can be both `require`d into CommonJS or `import`ed into an ES module.

Author: This section is just copied from our roadmap document.

The wording is careful to allow what you’re describing: an ESM-only package can still be “`require`d into CommonJS or `import`ed into an ES module” if `require` of ESM happens.

> - **Package path maps.** Along the lines of the browser [import maps proposal](https://github.com/WICG/import-maps), we would like to support paths to files within packages. This would allow things like `import sdk from 'some-service/sdk'` to have `'some-service/sdk'` map to something like `./src/sdk/index.mjs`.

Author: The proposal doesn’t mention bringing support to CommonJS, at least not yet. It hasn’t been updated in a while.
